client/app/views/account.coffee

```diff
@@ -108,7 +131,122 @@ module.exports = class exports.AccountView extends BaseView
+    on2faStatusChageSuccess: (message) =>
+        @info2faAlert.html t message
+        @info2faAlert.fadeIn()
+        clearTimeout hideFunc
+        hideFunc = setTimeout =>
+            window.location.reload()
+        , 2000
+    
+    
+    # Displays error on the two-factor auth panel.
+    # This one is also used for counter reset.
+    on2faError: =>
+        console.error err
+        @error2faAlert.fadeIn()
+        setTimeout =>
+            @error2faAlert.fadeOut()
+        , 5000
+            
+    
+    # Enables two-factor auth by generating a new key (if there's already one
+    # in base, resetting it), and set the HOTP counter if the auth type is HOTP
+    on2faEnableSubmit: (event) ->
+        authType = @twoFactorField.val()
+        
+        form =
+            authType: authType
+            otpKey: @getOtpKey()
+            hotpCounter: 0 if authType is 'hotp'
+        @twoFactorSubmitButton.spin true
+        
+        request.post 'api/user', form, (err, data) =>
+            @twoFactorSubmitButton.spin false
+            if err
+                @on2faError()
+            else
+                if data.success
+                    @on2faStatusChageSuccess 'account 2fa enabled'
+                else
+                    @on2faError()
+                    
+    
+    # Disables two-factor auth. No resetting of the key nor the counter, but
+    # that will happen if the user enables it back.
+    on2faDisableSubmit: (event) ->
+        form =
+            authType: null
+        
+        request.post 'api/user', form, (err, data) =>
+            @twoFactorSubmitButton.spin false
+            if err
+                @on2faError()
+            else
+                if data.success
+                    @on2faStatusChageSuccess 'account 2fa disabled'
+                else
+                    @on2faError()
+
 
+    # Resets the HOTP counter. This function can be called only in HOTP mode
+    # (and it will set the 2FA mode to HOTP otherwise).
+    # We don't use @on2faStatusChageSuccess here because we don't need to
+    # reload the page after resetting the counter.
+    on2faResetSubmit: (event) ->
+        form =
+            authType: 'hotp'
+            hotpCounter: 0
+        
+        request.post 'api/user', form, (err, data) =>
+            @twoFactorSubmitButton.spin false
+            if err
+                @on2faError()
+            else
+                if data.success
+                    @info2faAlert.html t 'account 2fa reset'
+                    @info2faAlert.fadeIn()
+                    clearTimeout hideFunc
+                    hideFunc = setTimeout =>
+                        @info2faAlert.fadeOut()
+                    , 5000
+                else
+                    @on2faError()
+
+
+    # OTP key is generated client-side, in this function. If consists in a
+    # 10-char long alphanumeric random string.
+    getOtpKey: () ->
+        buf = []
```

Comment:
We probably should prefer a PRNG rather than the `Math.random()` method to generate a cryptographically randomized string. To generate a hex representation of a 128bits hash (aka a 32chars hex string):
```coffee
# 2 chars per byte, 8 chars per PRNG, so 4 PRNG for a 128bits lengths
PRNG = new Uint32Array 4
window.crypto.getRandomValues PRNG

key = PRNG.reduce (key, subkey) ->
    # pad each PRNG hex with '0'
    key += ('0000' + subkey.toString(16)).slice(-8)
, ''
```